// fxs/src/main/js/index.js
/**
*	Default settings.
*/
exports.defaults = 
{
	host: "127.0.0.1",
	port: 843
}

/**
*	CrossDomainServer and load shortcut.
*
* 	@param {conf} Path to the configuration file.
*
* 	@api public
*/
exports.start = function( conf, policy )
{
	var cds = new exports.CrossDomainServer();
	cds.load( conf, policy );
	return cds;
}

/**
*	CrossDomainServer and load shortcut.
*
* 	@api public
*/
exports.server = function()
{
	var cds = new exports.CrossDomainServer();
	return cds;
}

/**
*	CrossDomainClient creation shortcut.
*
* 	@param {port} The port to send the policy file request to.
* 	@param {host} The host to send the policy file request to.
*
* 	@api public
*/
exports.client = function( port, host )
{
	var cdc = new exports.CrossDomainClient( port, host );
	return cdc;
}

/**
*	CrossDomainClient creation and ping shortcut.
*
* 	@param {port} The port to send the policy file request to.
* 	@param {host} The host to send the policy file request to.
*
* 	@api public
*/
exports.ping = function( port, host )
{
	var cdc = new exports.CrossDomainClient( port, host );
	if( !port )
	{		
		port = exports.defaults.port;
	}
	if( !host )
	{
		host = exports.defaults.host;
	}
	cdc.ping( port, host );
	return cdc;
}

/**
*	CrossDomainServer constructor.
*
*	@api public
*/
exports.CrossDomainServer = require('./xdomain-server');

/**
*	CrossDomainClient constructor.
*
*	@api public
*/
exports.CrossDomainClient = require('./xdomain-client');

/**
*	Version
*/
exports.version = '0.1.0';
